### DISCUSSION TITLE
Where do I find binaries?

### DISCUSSION THREAD

**Discussion by alandefreitas** [actions/virtual-environments#7620] [General questions] [answered]
Hi.

This might be a silly question but I have read the docs and it's a problem I'm still having nonetheless. 

I'm writing [many actions](https://github.com/alandefreitas/cpp-actions) that need to programmatically find and sometimes iterate the available software to find a compatible version for a workflow.
In the "Included Software" section I can see many tools I can use but I don't know where and, most importantly, how to programmatically find these executables.

When working on Ubuntu, things are slightly simpler because I can make some assumptions by replicating the APT installation process locally.
For instance, if I'm looking for `g++` >=11 <14, I assume they are installed with APT, list the APT packages installed, find a compatible version X.Y.Z, and look for the executable `g++-X` in `/usr/bin`. 
This is not ideal because it only works because I know conventions about how APT and GCC interact, but it still works.

On MacOS and Windows, I have absolutely no idea what to do. I often need to iterate the versions of `python` or `Mingw-w64` on Windows but have no idea where to find these executables. 

Do people just `find / ... python.exe` once to figure this out than assume a pattern? Sometimes I don't even know what executables to look for. What is there a general procedure to know where the executables are?

**Comment by vpolikarpov-akvelon** [comment, marked as answer]
Hi @alandefreitas.

All of the executables that you can use 'out of the box' are located in directories listed in PATH environment variable. I believe it's the easiest way to find all of them.

Also you can try to examine installation scripts. For example, scripts for Windows are here: https://github.com/actions/runner-images/tree/main/images/win/scripts/Installers

**Reply by soryy708** [reply]
Link is broken
